frame/staking/src/lib.rs

```diff
@@ -1846,50 +1864,52 @@ decl_module! {
 		///    minimized (to ensure less variance)
 		///
 		/// # <weight>
-		/// E: number of edges. m: size of winner committee. n: number of nominators. d: edge degree
-		/// (16 for now) v: number of on-chain validator candidates.
-		///
-		/// NOTE: given a solution which is reduced, we can enable a new check the ensure `|E| < n +
-		/// m`. We don't do this _yet_, but our offchain worker code executes it nonetheless.
-		///
-		/// major steps (all done in `check_and_replace_solution`):
-		///
-		/// - Storage: O(1) read `ElectionStatus`.
-		/// - Storage: O(1) read `PhragmenScore`.
-		/// - Storage: O(1) read `ValidatorCount`.
-		/// - Storage: O(1) length read from `SnapshotValidators`.
-		///
-		/// - Storage: O(v) reads of `AccountId` to fetch `snapshot_validators`.
-		/// - Memory: O(m) iterations to map winner index to validator id.
-		/// - Storage: O(n) reads `AccountId` to fetch `snapshot_nominators`.
-		/// - Memory: O(n + m) reads to map index to `AccountId` for un-compact.
-		///
-		/// - Storage: O(e) accountid reads from `Nomination` to read correct nominations.
-		/// - Storage: O(e) calls into `slashable_balance_of_vote_weight` to convert ratio to staked.
-		///
-		/// - Memory: build_support_map. O(e).
-		/// - Memory: evaluate_support: O(E).
-		///
-		/// - Storage: O(e) writes to `QueuedElected`.
-		/// - Storage: O(1) write to `QueuedScore`
-		///
-		/// The weight of this call is 1/10th of the blocks total weight.
+		/// All weight notes are pertaining to the case of a better solution, in which we execute the
+		/// longest code path.
+		/// Weight: 0 + 35 * v + 25 * n
+		/// State reads:
+		/// 	- Initial checks:
+		/// 		- ElectionState, CurrentEr, QueuedScore
+		/// 		- SnapshotValidators.len()
+		/// 		- ValidatorCount
+		/// 		- SnapshotValidators
+		/// 		- SnapshotNominators
+		/// 	- Iterate over nominators:
+		/// 		- compact.len() * (Nominators(who) + SlashingSpans(who)) (self vote does not have this, yet we cannot easily know)
```

Comment:
is it correct ?
it should be:
* for each nominators: Nominators
* for each validator nominated: SlashingSpans (also I think we check after this loop if the validator nominated is actually part of winners but anyway this is bound to 16)

Reply:
Most accurate is: For each assignment (which is a subset of nominators who are _active_ and is indeed same as `compact.len()`), we read `nominators`. If it is a validator, we don't read anything further. If it is nominator, we read the slashing span at most `distribution.len()` times (_I missed this!_ fixed it now).

Problem is that we cannot know if a vote is self vote or not. So we assume worse case and assume none are self vote. 

UPDATE: In principle, you could guess that 99.99% of the time and in any non-malicious solution, if you have `d` assignments in total, `winners.len()` of them are indeed self vote, since under the staking system, every validator votes for themselves. And the remaining or non-self-votes which need to read the slashing span.

For `distribution.len()`, we can either assume max and refund as we run the function, or take an average which can be at most 15 units in accurate. For now I've done the average, but open to change.

Reply:
I don't know what is better, I hardly see an attack there, but usually weight should always weight for the maximum malicious input, So instead of:
```rust
				.saturating_add( // SlashingSpans
					T::DbWeight::get().reads(
						compact.len()
							.saturating_sub(winners.len())
							.saturating_mul(compact.average_edge_count())
							as Weight
					)
				)
```
it should be `compact.edge_count()-winners.len()` but indeed this number can go very high. maybe the check could do 2 loops one to get the winners the other to check the assigments (and we first check it is parts of winners before decoding slashing spans).

But I guess if there is no possible attack here it's fine. though I can't tell.